### ISSUE TITLE
channel.sendToQueue returns no error if queue doesn't exist

### ISSUE THREAD

**Issue by stiofanoriain** [squaremo/amqp.node#633]
Hi all

Let's say the connection and channel are fine, but the queue has somehow deleted itself.

I use code like this to publish data to the queue:

    var sendData = function(data) {

      try {

        channel.sendToQueue(queue, new Buffer.from(JSON.stringify(data)), { persistent: true, mandatory: true }, function(err, ok) {

          if (err) console.warn(err);

        });

      } catch(err) {

        console.warn(err);

      }
    }

No error is returned. The data is sent to the queue and it acts as if everything is fine, however nothing has reached the queue as it does not exist.

Is there a known solution to this?

Thanks

**Comment by squaremo** [comment]
The mandatory flag will provoke the server into _returning_ the message as unroutable, if it can't be put in any queues.
(https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.publish.mandatory). This means you'll get a `'return'` event on the channel. Publishing is asynchronous, meaning the return value does not indicate anything about what happens at the server.

If you use [confirmations](https://www.rabbitmq.com/confirms.html#publisher-confirms), an unroutable message will still be acknowledged, but the returned message will be sent first, so you will be able to tell what happened. To use confirmations, obtain a channel with [`createConfirmChannel`](https://www.squaremobius.net/amqp.node/channel_api.html#model_createConfirmChannel).

**Comment by cressie176** [comment]
Hi @stiofanoriain,

I'm curious to know why you are reconnecting in the return handler. Just because a message was unrouteable doesn't necessarily mean the connection was dropped.

**Comment by stiofanoriain** [comment]
Hi @cressie176 

I've noticed if a minor network error causes the channel or connection or queue to go down, it doesn't recover, so my system won't be able to publish messages to the queue.

Rather than waiting to notice this, if there's an error I just automatically disconnect and reconnect everything.

While things are down I cache my messages and then publish them in one batch once the connection, channel and queue are back up.

Basically I need almost 100% uptime so I'm trying to automate reconnecting if there's a problem.

Is there a better solution?

Cheers

**Comment by cressie176** [comment]
Is your queue configured with auto-delete or exclusive? I can't think of any other reasons why the queue would be deleted on network error.

Instead if you can ensure the queue is durable and not auto-delete / exclusive, then you wouldn't have to recreate it. You will still have to automate connection and channel recovery for your producer. The consumer will need reconnect, establish a new channel and consume again. The connection and channel objects emit `error` and `close` events which help you detect this. The consumer message callback can also receive a `null` message if the broker shutsdown or the if queue is deleted / relocated.

https://www.cloudamqp.com/blog/part3-rabbitmq-best-practice-for-high-availability.html is worth a read too.
